gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSub.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final long serialVersionUID = -5220474819637439937L;
+
+    private final Option option;
+    private final Object value;
+
+    enum Option {
+      RETURN_IMMEDIATELY, MAX_MESSAGES
+    }
+
+    private PullOption(Option option, Object value) {
+      this.option = option;
+      this.value = value;
+    }
+
+    Option option() {
+      return option;
+    }
+
+    Object value() {
+      return value;
+    }
+
+    public static PullOption returnImmediatly() {
+      return new PullOption(Option.RETURN_IMMEDIATELY, true);
+    }
+
+    public static PullOption maxMessages(int maxMessages) {
+      return new PullOption(Option.MAX_MESSAGES, maxMessages);
+    }
+  }
+
+  /**
+   * A callback to process pulled messages.
+   * The message will be ack'ed upon successful return or nack'ed if exception is thrown.
+   */
+  interface PullCallback {
+    void process(Message message) throws Exception;
+  }
+
+  Topic create(TopicInfo topic);
+
+  Future<Topic> createAsync(TopicInfo topic);
+
+  // null if not found
+  Topic getTopic(String topic);
+
+  Future<Topic> getTopicAsync(String topic);
+
+  // false if not found
+  boolean deleteTopic(String topic);
+
+  Future<Boolean> deleteTopicAsync(String topic);
+
+  Page<Topic> listTopics(ListOption... options);
+
+  Future<AsyncPage<Topic>> listTopicsAsync(ListOption... options);
+
+  String publish(String topic, Message message);
+
+  Future<String> publishAsync(String topic, Message message);
+
+  List<String> publish(String topic, Message message, Message... messages);
+
+  Future<List<String>> publishAsync(String topic, Message message, Message... messages);
+
+  List<String> publish(String topic, Iterable<Message> messages);
+
+  Future<List<String>> publishAsync(String topic, Iterable<Message> messages);
+
+  Subscription create(SubscriptionInfo subscription);
+
+  Future<Subscription> createAsync(SubscriptionInfo subscription);
+
+  // null if not found
+  Subscription getSubscription(String subscription);
+
+  Future<Subscription> getSubscriptionAsync(String subscription);
+
+  void replacePushConfig(String subscription, PushConfig pushConfig);
+
+  Future<Void> replacePushConfigAsync(String subscription, PushConfig pushConfig);
+
+  // false if not found
+  boolean deleteSubscription(String subscription);
+
+  Future<Boolean> deleteSubscriptionAsync(String subscription);
+
+  Page<Subscription> listSubscriptions(ListOption... options);
+
+  Future<AsyncPage<Subscription>> listSubscriptionsAsync(ListOption... options);
+
+  Page<Subscription> listSubscriptions(String topic, ListOption... options);
+
+  Future<AsyncPage<Subscription>> listSubscriptionsAsync(String topic, ListOption... options);
+
+  List<ReceivedMessage> pull(String subscription, PullOption... options);
+
+  Future<List<ReceivedMessage>> pullAsync(String subscription, PullOption... options);
+
+  void pull(String subscription, PullCallback callback, PullOption... options);
```

Comment:
Are you suggesting keeping a synchronous pull with callback and to keep the 2 possible options for that call (and that call only)? Or are you suggesting to keep the synchronous pull and also leave the options for both synchronous and asynchronous methods?

The reason I didn't like keeping the synchronous pull is that in that case returning a MessageConsumer is strange. If it is synchronous than I think it is more expected that it should start pulling when called and not when calling start on its return value.

The reason I removed the options from the pull with callback flavor is to avoid strange/unexpected behavior such as:

1. no more messages will ever be sent to the callback but stop was not called (it is clear
that you reached the end when you get a list, it is not clear when you use callback).
2. no messages will ever sent as `return_immediately` was specified and there were no messages when `start()` was called.

Honestly, I think `return_immediately` is a strange beast and the Pub/Sub doc for "The client may cancel the request if it does not wish to wait any longer for the response." is vague (how do you do that for synchronous call...?)

In addition I think that path with the least user-surprises would be:

1. remove `return_immediately` from the API options (internally use `true` for the list pull and use `false` for the continuous callback).
2. only provide this pull related APIs:
    1.  list pull with an option for `max_messages`.
    2. async pull with callbacks with no options and pulling flow is only controlled by the returned `MessageConsumer`.

Note, that though the pull/fetch is async (and continuous) the calls to the `MessageProcesssor`
are sequential. We could (and I think that is a good idea) provide different settings to the pull with callback method such as max-buffer-size and max-number of concurrent callback calls (or this could be set on the `MessageConsumer` and documented that it takes affect only when called before `start`).

I think it would be OK to also have a synchronous pull with callback that is triggered immediately
can specify `max_messages` and, as with the list pull, will internally use `true` to `return_immediately`. Also, for simplicity I think this method should always process the messages sequentially. However, though easy to add, I would not rush with implementing it
as it would not be easy to remove if we decided that it is just confusing or inconsistent.

Thoughts?  /cc @eschapira